### ISSUE TITLE
500 error on file search

### ISSUE THREAD

**Issue by zachmullen** [girder/girder#3181]
When executing a search via the API on the `file` collection, we get an exception thrown indicating that the sort operation is not able to use an index.

The exception report is [here](https://sentry.io/organizations/kitware-data/issues/1448285819/?project=1474412).

Request to reproduce:

`curl -X GET --header 'Accept: application/json' 'https://data.kitware.com/api/v1/resource/search?q=test.png&mode=text&types=%5B%22file%22%5D&limit=10'`

**Comment by zachmullen** [comment]
The top of the stack trace exported from Sentry:

```
OperationFailure: Executor error during find command :: caused by :: Sort operation used more than the maximum 33554432 bytes of RAM. Add an index, or specify a smaller limit.
(12 additional frame(s) were not displayed)
...
  File "pymongo/mongo_client.py", line 1347, in _run_operation_with_response
    exhaust=exhaust)
  File "pymongo/mongo_client.py", line 1465, in _retryable_read
    return func(session, server, sock_info, slave_ok)
  File "pymongo/mongo_client.py", line 1341, in _cmd
    unpack_res)
  File "pymongo/server.py", line 136, in run_operation_with_response
    _check_command_response(first)
  File "pymongo/helpers.py", line 159, in _check_command_response
    raise OperationFailure(msg % errmsg, code, response)

500 Error
```

**Comment by zachmullen** [comment]
I believe I now understand the cause of this. It's happening inside `AclMixin.textSearch()` on the `File` model.

For the `File` model, we have to fallback on the old `filterResultsByPermission` behavior wherein we do the cursor filtering in the app rather than the database. Because we are passing a sort field (the `$meta: 'textScore'` value) without providing a sane `limit` value, it can blow up.

One option would be to not do textScore sorting on File text searches and just display them in arbitrary order. Another option would be to pass some sensible limit that is larger than the requested limit to allow for filtering, but not so large that it could blow up the database by sorting. A robust solution would be to request batches of a fixed size until we exhaust the values or fill the filtered results list.

ping @manthey , any thoughts?

**Comment by manthey** [comment]
Part of the issue here is that searching for `test.png` will match `test` or `png`, so the result set is large.  The user probably wanted to do a search for `"test.png"` where the filename has to include that text as a case insensitive substring.  Returning unsorted results is lousy.

When you use an aggregation, you can tell mongo to use the disk if the results are too big for memory, but I don't think you can do that for a cursor sort.  Since sorted results are really valuable, I see two user-friendly solutions: implement the multiple-hop aggregation for non-asset-controlled models that are asset controlled by grand-parents, or try as we do and when there is an exception, then (and only then) try again without the sort.

**Comment by zachmullen** [comment]
How much work do you think it would be to implement the two-hop aggregation pipeline? Also, can you comment on this proposal?

> request batches of a fixed size until we exhaust the values or fill the filtered results list

**Comment by manthey** [comment]
Batches won't fix the problem with exhausting memory for sorting.  The sort has to happen before it can determine what to return in a batch.

**Comment by manthey** [comment]
I think if we are going to do it, we should write the *n*-hop aggregation, rather than just the two-hop.  It probably would be a few hours of work, as I relatively recently have made some complex aggregations.
